### PR TITLE
[#117991263] upgrade concourse to 1.5.1

### PR DESCRIPTION
**Pull request by mtekel**
## What

[Upgrade Deployer Concourse](https://www.pivotaltracker.com/n/projects/1275640/stories/117991263)

## How to review

Run bootstrap pipeline & watch it upgrade your concourse. Run your new concourse to confirm all still works fine. A good test is to run pipeline against existing environment and observe if anything is being changed.

## Merging
Before you merge check if any other PRs have been merged since this PR was opened. If so, check if any changes to pipelines use/introduce `image:` definition of container. If so, rebase on master and convert this to `image_resource` like the rest of container specifications in the pipelines. Note that bootstrap create/destroy and auto-delete pipelines should still use `image:` as they are meant for older bootstrap concourse.

This PR depends on https://github.com/alphagov/paas-docker-cloudfoundry-tools/pull/60. Review that as a part of reviewing this PR. Remove 922fec7 after https://github.com/alphagov/paas-docker-cloudfoundry-tools/pull/60 is merged.

## Who can review

not @combor @mtekel or @benhyland

### CODE REVIEW

**Comment by jimconner** [comment]
It is not possible for outsiders to view the 'What' on this story as we don't all have access to the private tracker used.

**Comment by mtekel** [comment]
That's a pity. It links to an epic story of concourse upgrade where we attempted to upgrade to version 1.0 at first, finding out we couldn't. We made a PR to fix the issues for bosh-init, which were merged to 1.3 concourse, which unfortunately had issue with too many containers. Then we tried 1.4.0 and 1.4.1, which still didn't fix the issue. Finally 1.5.1 seems to work OK. Except we now need to extend fly hijack to be able to specify container type to hijack, so that we don't get multiple container choices when doing bosh-cli...

File: concourse/pipelines/create-bosh-cloudfoundry.yml [hunk 1]
```diff
@@ -1814,14 +1969,17 @@ jobs:
             - -c
             - |
               VAL_FROM_YAML=$(pwd)/paas-cf/concourse/scripts/val_from_yaml.rb
-              CONCOURSE_ATC_PASSWORD=$("$VAL_FROM_YAML" jobs.concourse.properties.atc.basic_auth_password concourse-manifest/concourse-manifest.yml)
+              CONCOURSE_ATC_PASSWORD=$("$VAL_FROM_YAML" jobs.concourse.properties.basic_auth_password concourse-manifest/concourse-manifest.yml)
               export CONCOURSE_ATC_PASSWORD
               echo Looking for non running VMs
-              ./paas-cf/concourse/scripts/bosh-cli.sh bosh status
+              echo 2 | ./paas-cf/concourse/scripts/bosh-cli.sh bosh status
```
Comment (dcarley) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: Can we specify `fly intercept -c2 …` in `bosh-cli.sh` instead of using `echo` here and documenting in the README?

Comment (mtekel) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: `error: invalid argument for flag `-c, --check' (expected flaghelpers.ResourceFlag): argument format should be <pipeline>/<resource>`

Comment (dcarley) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: Does using `<pipeline>/<resource>` work? Using a name would be more clear and reliable than using a number.

Comment (dcarley) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: If that doesn't work, then it might be `-s2` that we can use.

Comment (mtekel) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: But I don't want to hijack resource, but a job. run-bosh-cli is not a name of resource...

Comment (mtekel) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: -s2 doesn't work. '2' is not a name of step. I only get these options in the fly:
```
[hijack command options]
      -j, --job=PIPELINE/JOB        Name of a job to hijack
      -c, --check=PIPELINE/CHECK    Name of a resource's checking container to hijack
      -b, --build=                  Build number within the job, or global build ID
      -s, --step=                   Name of step to hijack (e.g. build, unit, resource name)
      -a, --attempt=                Attempt number of step to hijack. Can be specified multiple times for nested retries
```
I wish I did not have to do this but I did not find a way to tell fly that I want to hijack a 'task' type container instead of 'check':
```
1: build #20, step: run-bosh-cli, type: check
2: build #20, step: run-bosh-cli, type: task
```
These 2 have same pipeline name, job name and step name. You can't currently specify any further details in fly.

**Comment by keymon** [comment]
Deploys good for me... lets see how it deploys :)

**Comment by mtekel** [comment]
It seems like t2.medium is not fast enough for this version/setup of concourse. I left mine running over the weekend doing nothing. Initially it was regaining credits, but then it stopped next day. Current load is 0.4, which is exactly how much credits that instance gets. Looking at the server, there's not much going on. It is possible that if e.g. there is some issue/change with external dependencies (dockerhub, github), resource checks can consume more CPU. I suggest we play [Change deployer-concourse to use a non-burstable instance type.](https://www.pivotaltracker.com/n/projects/1275640/stories/126234443) story. c4.large looks like a good replacement or m4.large if we need 8G of RAM (for e.g. disk cache).

**Comment by dcarley** [comment]
@mtekel thanks for checking. Is the new version usable on the existing instance type? Will we be able to upgrade first and then change the instance type afterwards, in separate stories?

@bleach any concerns about the cost? Urmi won't be about to ask.

**Comment by mtekel** [comment]
Yes we'll be able to do 2 changes, but it means 2 changes (someone has to manually run bootstrap pipeline in all environments for each of the stories). 

Regarding costs: c4.large is 2x more expensive than t2.medium. m4.large is 2.5x times more expensive. For our ~ 12 environments (dev + ci + prod + stage), the difference per year was about $5k (we checked this with @benhyland on Friday).

**Comment by alext** [comment]
For reference, a t2.medium costs $0.056/hr and a c4.large costs $0.119/hr

**Comment by keymon** [comment]
Can we explain why we cannot use `link()` in 4f236663c15844f8ab18e2f8614897145211b6a7

Something like:

> `bosh_init` does not support [link services as bosh does](https://bosh.io/docs/links.html). Concourse release now relies on links to configure services between jobs, although it can be disabled specifying all the properties. This can break in the future, as concourse does not recommend use `bosh_init` to deploy it.

**Comment by keymon** [comment]
I would squash b0c38ae1566c38e7d7fb7689de747f2a7accb6c4 and23375c703a27daf7cc94c5272dfbd7047c67cd14, and also explain that we need to keep `image` in the job definition until https://github.com/concourse/fly/issues/98 is fixed.

**Comment by keymon** [comment]
The PR seems fine, concourse deploys and the pipeline runs successfully.

We need to clarify if we want to upgrade the instance type or not.

**Comment by bleach** [comment]
I think the cost of upgrading to a c4.large is justified if the current instance size causes our concourses to effectively and the larger size fixes that problem without us spending too much engineer time on it. It seems like this is the case from the comments above.

If this problem is introduced or made much worse by the upgrade then I would opt to fix it during the upgrade; if it does cause problems it will hamper our ability to get anything delivered to production.

**Comment by bleach** [comment]
tl;dr; if the choice is between between paying $5k/ year to upgrade all our concourses or having them grind to a halt every so often, we should pay the $5k

**Comment by mtekel** [comment]
Given the situation (and the fact that it keeps breaking), I'd say changing instance type is not much of a choice. If we care about the costs, we can possibly save more elsewhere (e.g. different VM types in cloud config for dev).

I have updated the comment messages, squashed commits and PR description. 
922fec7 should be removed after https://github.com/alphagov/paas-docker-cloudfoundry-tools/pull/60 is merged.

**Comment by dcarley** [comment]
@mtekel thanks for finding a solution to the bosh-cli usage and let's go ahead with the instance type change 👍

**Comment by mtekel** [comment]
half of the credit for the solution goes to @keymon

**Comment by keymon** [comment]
LGTM, merging!

**Comment by HenryTK** [comment]
Now that #364 has been merged there is a reference to `image` used instead of `image_resource`, in the [`update-kibana-timezone` task](https://github.com/alphagov/paas-cf/blob/master/concourse/pipelines/create-bosh-cloudfoundry.yml#L1520), so this branch needs updating before merging.

**Comment by keymon** [comment]
> Now that #364 has been merged there is a reference to image used instead of image_resource, in the update-kibana-timezone task, so this branch needs updating before merging.  

Rebased and updated

File: concourse/pipelines/create-bosh-cloudfoundry.yml [hunk 2]
```diff
@@ -1402,7 +1527,11 @@ jobs:
         - task: sync-admin-users
           config:
             platform: linux
-            image: docker:///ruby#2.2
+            image_resource:
+              type: docker-image
+              source:
+                repository: ruby
+                tag: 2.2-slim
```
Comment (alext) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: This changed the container from `2.2` to `2.2-slim`. This has led to errors running the sync-admin-users script as it can no longer `bundle install` because the slim container doesn't include a compiler.

I'll do a PR to fix this.
